database/src/table.rs

```diff
@@ -1,181 +1,49 @@
-//! ### Table module
-//! This module contains the definition of the [`Table`] and [`DupTable`] trait, and the actual tables used in the database.
-//! [`DupTable`] are just a trait used to define that they support DUPSORT|DUPFIXED operation (as of now we don't know the equivalent for HSE).
-//! All tables are defined with docs explaining its purpose, what types are the key and data.
-//! For more details please look at Cuprate's book : <link to cuprate book>
-
-use crate::{
-    encoding::Compat,
-    types::{
-        /*OutTx,*/ AltBlock, BlockMetadata, /*RctOutkey,*/ OutputMetadata,
-        TransactionPruned, TxIndex, /*OutAmountIdx,*/ /*KeyImage,*/ TxOutputIdx,
-    },
-};
-use bincode::{de::Decode, enc::Encode};
-use monero::{blockdata::transaction::KeyImage, Block, Hash};
-
-/// A trait implementing a table interaction for the database. It is implemented to an empty struct to specify the name and table's associated types. These associated
-/// types are used to simplify deserialization process.
-pub trait Table: Send + Sync + 'static + Clone {
-    // name of the table
-    const TABLE_NAME: &'static str;
-
-    // Definition of a key & value types of the database
-    type Key: Encode + Decode;
-    type Value: Encode + Decode;
-}
-
-/// A trait implementing a table with duplicated data support.
-pub trait DupTable: Table {
-    // Subkey of the table (prefix of the data)
-    type SubKey: Encode + Decode;
-}
-
-/// This declarative macro declare a new empty struct and impl the specified name, and corresponding types.
-macro_rules! impl_table {
-	( $(#[$docs:meta])* $table:ident , $key:ty , $value:ty ) => {
-        #[derive(Clone)]
-		$(#[$docs])*
-		pub(crate) struct $table;
-
-   		impl Table for $table {
-	 		const TABLE_NAME: &'static str = stringify!($table);
-        	type Key = $key;
-            type Value = $value;
+//! Database table abstraction; `trait Table`.
+
+//---------------------------------------------------------------------------------------------------- Import
+use crate::pod::Pod;
+
+//---------------------------------------------------------------------------------------------------- Table
+/// Database table metadata.
+///
+/// Purely compile time information for database tables.
+/// Not really an accurate name for `K/V` database but
+/// this represents the metadata of a `K/V` storing object.
+pub trait Table {
+    // TODO:
+    //
+    // Add K/V comparison `type`s that define
+    // how this table will be stored.
+    //
+    // type KeyComparator: fn(&Self::Key, &Self::Key) -> Ordering;
+    // type ValueComparator: fn(&Self::Value, &Self::Value) -> Ordering;
+
+    /// Name of the database table.
+    const NAME: &'static str;
+
+    cfg_if::cfg_if! {
+        if #[cfg(all(feature = "sanakirja", not(feature = "heed")))] {
+            // TODO: fix this sanakirja bound.
+
+            /// Concrete key type.
+            type Key: Pod + sanakirja::Storable;
```

Comment:
I think we should have a trait for DBKeys as well: 
```rust
trait DBKey {
    /// A bool for if this table requires multiple keys to reach a value.
    const DUP: bool;
    type Primary;
    /// This will only be used for DUP tables so can be set to a default type for single key tables
    type Secondary;

    fn primary(self) -> Self::Primary;

    fn primary_secondary(self) -> (Self::Primary, Self::Secondary);
}
```

So then the back-end would do something like:

```rust
fn get_val<T: Table>(key: T::Key) -> T::Value {
   if <T::Key as DBKey>::DUP {
      let (primary, secondary) = key.primary_secondary();
      // get value with both keys
   } else {
       // get val normally.
   }
}
```

Reply:
Should this be on `trait Table`? It would then be associated with a table:
```rust
pub trait Table {
    const NAME: &'static str;
    const CONSTANT_SIZE: bool;
    const DUPLICATE: bool;

    type KeyPrimary: Pod;
    type KeySecondary: Pod;
    type Value: Pod;

    fn primary(self) -> Self::Primary;
    fn primary_secondary(self) -> (Self::Primary, Self::Secondary);
}
```
If it were a separate trait, it would be implemented on the key type itself so 2 tables with the same key would conflict, right?
```rust
trait DBKey { /* ... */ }
impl DBKey for u8 {
    /* ... */
    const DUP: bool = true;
}

// Table with dup keys.
impl Table for Table1 {
    /* ... */
    type Key = u8;
}

// Table without dup keys.
impl Table for Table2 {
    /* ... */
    type Key = u8; // <- same key, different table, DBKey impl conflicts
}
```

Reply:
It could be, if you think that would be better.

My idea was for a type like `u8` to always be `DUP: bool = false` and for there to be a single type `DupKey` that sets `DUP = true`. The `DupKey` would look something like:

```rust
struct DupKey<P, S> {
   primary: P,
   secondary: S 
}
```

and then if the table needed a dup key we would set the key to `DupKey<u8, u8>` for example. 

I am happy to do it either way.

Reply:
How would `fn primary(self) -> Self::Primary` work? `Table` is implemented on an empty struct so `self` is pretty much `()`.

Would this work?

```rust
struct DupKey<P, S> {
   primary: P,
   secondary: S 
}

trait Table {
    /* ... */
    fn primary_secondary(key: DupKey<Self::Primary, Self::Secondary>) -> (Self::Primary, Self::Secondary);
}

impl Table for /* table */ {
    type Primary = u8;
    type Secondary = u64;

    /* ... */

    fn primary_secondary(key: DupKey<u8, u64>) -> (Self::Primary, Self::Secondary) {
        (key.primary, key.secondary)
    }
}
```

Reply:
Although `DupKey` would have to be the `Primary` as well, so maybe:

```rust

impl Table for /* table */ {
    type Primary = DupKey<u8, u64>;
    type Secondary = u64;

    /* ... */

    fn primary_secondary(key: DupKey<u8, u64>) -> (Self::Primary, Self::Secondary) {
        (key, key.secondary)
    }
}
```
is `fn primary()` needed? We would have direct access to `T::Key` anyway in the `fn get_val<T: Table>(key: T::Key) -> T::Value` example

Reply:
You are slowly getting to my idea :) I think it would probably be easier to go with the separate trait right?